int-set/src/bitset.rs

```diff
@@ -124,51 +144,88 @@ impl<T> BitSet<T> {
 
     /// Return the major value (top 23 bits) of the page associated with value.
     fn get_major_value(&self, value: u32) -> u32 {
-        return value >> PAGE_BITS_LOG_2;
+        value >> PAGE_BITS_LOG_2
     }
 
     fn major_start(&self, major: u32) -> u32 {
-        return major << PAGE_BITS_LOG_2;
+        major << PAGE_BITS_LOG_2
     }
 
-    /// Return a reference to the page that 'value' resides in.
-    fn page_for(&self, value: u32) -> Option<&BitPage> {
-        let major_value = self.get_major_value(value);
+    /// Returns the index in self.pages (if it exists) for the page with the same major as major_value.
+    fn page_index_for_major(&self, major_value: u32) -> Option<usize> {
         self.page_map
             .binary_search_by(|probe| probe.major_value.cmp(&major_value))
             .ok()
-            .and_then(|info_idx| {
-                let real_idx = self.page_map[info_idx].index as usize;
-                self.pages.get(real_idx)
-            })
+            .map(|info_idx| self.page_map[info_idx].index as usize)
+    }
+
+    /// Returns the index in self.pages for the page with the same major as major_value. Will create the page
+    /// if it does not yet exist.
+    fn ensure_page_index_for_major(&mut self, major_value: u32) -> usize {
+        match self
+            .page_map
+            .binary_search_by(|probe| probe.major_value.cmp(&major_value))
+        {
+            Ok(map_index) => self.page_map[map_index].index as usize,
+            Err(map_index_to_insert) => {
+                let page_index = self.pages.len();
+                self.pages.push(BitPage::new_zeroes());
+                let new_info = PageInfo {
+                    index: page_index as u32,
+                    major_value,
+                };
+                self.page_map.insert(map_index_to_insert, new_info);
+                page_index
+            }
+        }
+    }
+
+    /// Return a reference to the page that 'value' resides in.
+    fn page_for(&self, value: u32) -> Option<&BitPage> {
+        let major_value = self.get_major_value(value);
+        let pages_index = self.page_index_for_major(major_value)?;
+        Some(&self.pages[pages_index])
```

Comment:
I'd just write this as `self.pages.get(page_index)`; in general I think it's a good habit to avoid direct indexing, which often generates worse code (specifically the compiler has to generate a panic path for when indexing fails) 🤷

Reply:
Sure I'll give that a try and see if it effects the benchmarks in either direction.

Reply:
https://github.com/googlefonts/fontations/pull/896

Reply:
Bench times were neutral